File: packages/sb-uhq/src/self/commands/gestion/antigroup.js
"use strict";

const fs   = require("fs");
const path = require("path");
const { dataPath } = require("../../func/data-path");

const ANTIGROUP_FILE = dataPath("config", "antigroup.json");

// ── I/O ──────────────────────────────────────────────────────────────────────

function getState() {
  try {
    if (!fs.existsSync(ANTIGROUP_FILE)) return false;
    return JSON.parse(fs.readFileSync(ANTIGROUP_FILE, "utf-8")).enabled === true;
  } catch { return false; }
}

function setState(enabled) {
  fs.mkdirSync(path.dirname(ANTIGROUP_FILE), { recursive: true });
  fs.writeFileSync(ANTIGROUP_FILE, JSON.stringify({ enabled }, null, 2));
}

// ── Event handler Discord ─────────────────────────────────────────────────────

async function handleChannelCreate(client, channel) {
  if (channel.type !== "GROUP_DM" || channel.ownerId === client.user.id || !getState()) return;
  console.log(`[ANTIGROUP] Groupe DM détecté (${channel.id}), je quitte.`);
  channel.delete().catch(() => {});
}

// ── Logique pure (bridge uniquement) ─────────────────────────────────────────

/**
 * @param {import("discord.js-selfbot-v13").Client} client
 * @param {{ action: "toggle"|"getState"|"leaveAll" }} payload
 */
async function execute(client, payload) {
  const { action } = payload;

  if (action === "getState") return { enabled: getState() };

  if (action === "toggle") {
    const newState = !getState();
    setState(newState);
    return { enabled: newState };
  }

  if (action === "leaveAll") {
    // Fetch tous les channels depuis l'API pour ne pas se limiter au cache
    // (le cache ne contient que les channels actifs depuis le démarrage)
    await client.channels.fetch().catch(() => {});
    const groups = [...client.channels.cache.values()].filter(
      c => c.type === "GROUP_DM"
    );

    let left  = 0;
    let failed = 0;
    const details = [];

    for (const group of groups) {
      try {
        const name = group.name ?? `Groupe (${group.id})`;
        await group.delete();
        left++;
        console.log(`[ANTIGROUP] Quitté le groupe "${name}" (${group.id})`);
        details.push({ id: group.id, name, success: true });
      } catch (err) {
        failed++;
        console.error(`[ANTIGROUP] Impossible de quitter le groupe ${group.id} : ${err.message}`);
        details.push({ id: group.id, name: group.name ?? group.id, success: false, error: err.message });
      }
      // Petit délai pour éviter le rate-limit
      await new Promise(r => setTimeout(r, 300));
    }

    return { left, failed, total: groups.length, details };
  }

  throw new Error(`Action antigroup inconnue : '${action}'`);
}

module.exports = { name: "antigroup", execute, handleChannelCreate };

File: packages/sb-uhq/app/commands/gestion/antigroup.py
"""antigroup — quitte automatiquement les groupes DM entrants (+ leaveAll)."""

from __future__ import annotations

import asyncio

import discord

from ...func.data_path import data_path, read_json, write_json
from ...func.logbus import log, logerr

ANTIGROUP_FILE = data_path("config", "antigroup.json")


def get_state() -> bool:
    if not ANTIGROUP_FILE.exists():
        return False
    data = read_json(ANTIGROUP_FILE, {})
    return data.get("enabled") is True


def set_state(enabled: bool) -> None:
    write_json(ANTIGROUP_FILE, {"enabled": enabled})


def _is_group(channel) -> bool:
    return isinstance(channel, discord.GroupChannel) or getattr(channel, "type", None) == discord.ChannelType.group


async def handle_channel_create(client, channel) -> None:
    if not _is_group(channel):
        return
    if getattr(channel, "owner_id", None) == client.user.id:
        return
    if not get_state():
        return
    log(f"[ANTIGROUP] Groupe DM détecté ({channel.id}), je quitte.")
    try:
        await channel.leave()
    except Exception:
        pass


async def execute(client, payload):
    action = payload.get("action")

    if action == "getState":
        return {"enabled": get_state()}

    if action == "toggle":
        new_state = not get_state()
        set_state(new_state)
        return {"enabled": new_state}

    if action == "leaveAll":
        groups = [c for c in client.private_channels if _is_group(c)]
        left = 0
        failed = 0
        details = []
        for group in groups:
            name = getattr(group, "name", None) or f"Groupe ({group.id})"
            try:
                await group.leave()
                left += 1
                log(f'[ANTIGROUP] Quitté le groupe "{name}" ({group.id})')
                details.append({"id": str(group.id), "name": name, "success": True})
            except Exception as err:  # noqa: BLE001
                failed += 1
                logerr(f"[ANTIGROUP] Impossible de quitter le groupe {group.id} : {err}")
                details.append({"id": str(group.id), "name": name, "success": False,
                                "error": str(err)})
            await asyncio.sleep(0.3)
        return {"left": left, "failed": failed, "total": len(groups), "details": details}

    raise ValueError(f"Action antigroup inconnue : '{action}'")
